mixer/pkg/runtime/dispatcher.go

```diff
@@ -281,8 +280,40 @@ func (m *dispatcher) Quota(ctx context.Context, requestBag attribute.Bag,
 // Preprocess runs the first phase of adapter processing before any other adapters are run.
 // Attribute producing adapters are run in this phase.
 func (m *dispatcher) Preprocess(ctx context.Context, requestBag attribute.Bag, responseBag *attribute.MutableBag) error {
-	// FIXME
-	return errors.New("not implemented")
+
+	// protect the mutable responseBag
+	var lock sync.Mutex
+
+	_, err := m.dispatch(ctx, requestBag, adptTmpl.TEMPLATE_VARIETY_ATTRIBUTE_GENERATOR,
+		func(call *Action) []dispatchFn {
+			ra := make([]dispatchFn, 0, len(call.instanceConfig))
+			for _, inst := range call.instanceConfig {
+				ra = append(ra,
+					func(ctx context.Context) *result {
+						mBag, err := call.processor.ProcessGenAttrs(ctx, inst.Name,
+							inst.Params.(proto.Message),
+							requestBag, m.mapper,
+							call.handler)
+						if err == nil {
+							lock.Lock()
+							err = responseBag.Merge(mBag)
```

Comment:
If responseBag.merge panics, safeDispatch will end the goroutine without a proper cleanup.
This means the local `lock` is held by a goroutine that is effectively dead. At this point all other `lock.Lock` will block indefinitely.

need to `defer lock.Unlock()` which ensures cleanup. The cost is in nanosecs.